torch/nested/_internal/cached_tensor.py

```diff
@@ -0,0 +1,125 @@
+from contextlib import contextmanager
+from typing import *  # noqa: F403
+
+import torch
+
+
+class CachedTensor(torch.Tensor):
+    metadata: Dict[str, Optional[torch.Tensor]]
+    source_field: str
+
+    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
+    # is determined by a "source" tensor in the dict (specified by the user
+    # during construction).
+    #
+    # This tensor does not support any operations on it by default.
+    # To leverage the extra metadata, you must register an op to perform the special logic
+    # you want via register_cached_tensor_func.
+    #
+    # When used this way (1) it is a convenient way to keep around metadata
+    # related to a tensor, without having to laboriously thread those extra metadata
+    # around, e.g. through aten signatures. (2) allows one to trigger custom
+    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
+    # otherwise because the subclass constructor op itself usually does not take the
+    # subclass itself as input!). See NestedTensor for an example.
+
+    @staticmethod
+    @torch._disable_dynamo  # type: ignore[misc]
+    def __new__(
+        cls,
+        metadata: Dict[str, Optional[torch.Tensor]],
+        source_field: str,
+    ) -> "CachedTensor":
+        source = metadata.get(source_field)
+        assert source is not None
+        shape = source.shape
+        kwargs = {}
+        kwargs["strides"] = source.stride()
+        kwargs["storage_offset"] = source.storage_offset()  # type: ignore[assignment]
+        kwargs["device"] = source.device  # type: ignore[assignment]
+        kwargs["layout"] = source.layout  # type: ignore[assignment]
+        kwargs["requires_grad"] = source.requires_grad  # type: ignore[assignment]
+        kwargs["dtype"] = source.dtype  # type: ignore[assignment]
+        out = torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+        out.metadata = metadata
+        out.source_field = source_field
+        return out
+
+    def __repr__(self) -> str:  # type: ignore[override]
+        return f"CachedTensor({repr(self.metadata[self.source_field])})"
+
+    def __getattr__(self, name: str) -> Optional[torch.Tensor]:
```

Comment:
Is this custom getattr mainly to make sure that `cache_tensor.key` secretly maps to `cache_tensor.metadata.key`, so the source information of inner tensors is accurate?

That makes sense, although I'm not sure how good dynamo's handling of custom getattr's is. Do you ever expect attribute accesses on a `CachedTensor` to happen from user python code, that dynamo needs to inline?

Reply:
> Is this custom getattr mainly to make sure that cache_tensor.key secretly maps to cache_tensor.metadata.key, so the source information of inner tensors is accurate?

Yeah, this was mainly to get `__tensor_flatten__` to work

> That makes sense, although I'm not sure how good dynamo's handling of custom getattr's is. Do you ever expect attribute accesses on a CachedTensor to happen from user python code, that dynamo needs to inline?

Nah, today I expect CachedTensor to be purely private.